cli/cmd/common/utils.go

```diff
@@ -1,20 +1,81 @@
 package common
 
 import (
+	"embed"
 	"fmt"
 	"os"
 	"strings"
 
 	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart/loader"
 	helmCLI "helm.sh/helm/v3/pkg/cli"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 )
 
 const (
 	DefaultReleaseName      = "consul"
 	DefaultReleaseNamespace = "consul"
+	chartFileName           = "Chart.yaml"
+	valuesFileName          = "values.yaml"
+	templatesDirName        = "templates"
+	TopLevelChartDirName    = "consul"
 )
 
+// ReadChartFiles reads the chart files from the embedded file system, and loads their contents into
+// []*loader.BufferedFile. This is a format that the Helm Go SDK functions can read from to create a chart to install
+// from. The names of these files are important, as there are case statements in the Helm Go SDK looking for files named
+// "Chart.yaml" or "templates/<templatename>.yaml", which is why even though the embedded file system has them named
+// "consul/Chart.yaml" we have to strip the "consul" prefix out.
+func ReadChartFiles(chart embed.FS, chartDirName string) ([]*loader.BufferedFile, error) {
+	var chartFiles []*loader.BufferedFile
+
+	bytes, err := chart.ReadFile(fmt.Sprintf("%s/%s", chartDirName, chartFileName))
+	if err != nil {
+		return []*loader.BufferedFile{}, err
+	}
+	chartFiles = append(chartFiles,
+		&loader.BufferedFile{
+			Name: chartFileName,
+			Data: bytes,
+		},
+	)
+
+	bytes, err = chart.ReadFile(fmt.Sprintf("%s/%s", chartDirName, valuesFileName))
+	if err != nil {
+		return []*loader.BufferedFile{}, err
+	}
+	chartFiles = append(chartFiles,
+		&loader.BufferedFile{
+			Name: valuesFileName,
+			Data: bytes,
+		},
+	)
```

Comment:
You could wrap this behavior in a separate function to reduce code duplication.

```go
func readChart(chartDirName, fileName) (*chartFile, error) {
        bytes, err = chart.ReadFile(fmt.Sprintf("%s/%s", chartDirName, fileName))
	if err != nil {
		return nil, err
	}
	return &loader.BufferedFile{
		Name: fileName,
		Data: bytes,
	}
}
```

Reply:
I took a stab at a fun refactor using Thomas's suggestion and using `filepath.Rel` to do the removal of `charts/`:


```go
func ReadChartFiles(chart embed.FS, chartDirName string) ([]*loader.BufferedFile, error) {
	var chartFiles []*loader.BufferedFile

	// Load Chart.yaml and values.yaml first.
	for _, f := range []string{chartFileName, valuesFileName} {
		buffered, err := readFile(chart, filepath.Join(chartDirName, f), chartDirName)
		if err != nil {
			return nil, err
		}
		chartFiles = append(chartFiles, buffered)
	}

	// Now load everything under templates/.
	dirs, err := chart.ReadDir(filepath.Join(chartDirName, templatesDirName))
	if err != nil {
		return nil, err
	}
	for _, dir := range dirs {
		if dir.IsDir() {
			// We only need to package up files.
			continue
		}

		buffered, err := readFile(chart, filepath.Join(chartDirName, templatesDirName, dir.Name()), chartDirName)
		if err != nil {
			return nil, err
		}
		chartFiles = append(chartFiles, buffered)
	}

	return chartFiles, nil
}

func readFile(chart embed.FS, f string, basePath string) (*loader.BufferedFile, error) {
	bytes, err := chart.ReadFile(f)
	if err != nil {
		return nil, err
	}
	rel, err := filepath.Rel(basePath, f)
	if err != nil {
		return nil, err
	}
	return &loader.BufferedFile{
		Name: rel,
		Data: bytes,
	}, nil
}

```